### PR TITLE
Moment map continuum subtraction

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request implements support for continuum subtraction within the moment map plugin in cubeviz.  It does so by refactoring code in line-analysis into a reusable component, showing the same input options and visualization (but with additional descriptive text to explain that the continuum is re-computed per-spaxel when computing the moment map).

https://github.com/spacetelescope/jdaviz/assets/877591/012d84cf-a7a9-4a12-ad54-53b18387146d

This PR also includes a commit (since reverted but left in the PR diff in case we want to revisit it) which would enable exporting the continuum as a cube back into the app.

**TODO**:
- [x] proper deprecation of `line_analysis.width` -> `line_analysis.continuum_width`
- [x] investigate need to handle `orig_spec`
- [x] consider moving mark visibility logic to mixin

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by javerbukh** [comment]
Trigger traceback with the following steps:

1. Cubeviz example notebook
2. Create spectral subset
3. Go to moment map plugin
4. Spectral region is subset 1
5. Continuum surrounding
6. Moment 2 and wavelength
7. Plot in uncertainty viewer
8. Press calculate

Here is the traceback:
<details>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~\anaconda3\envs\test-mm-sub\lib\site-packages\ipyvue\VueTemplateWidget.py:60, in Events._handle_event(self, _, content, buffers)
     58     getattr(self, "vue_" + event)(data, buffers)
     59 else:
---> 60     getattr(self, "vue_" + event)(data)

File ~\Documents\STScI\jdaviz\jdaviz\configs\cubeviz\plugins\moment_maps\moment_maps.py:228, in MomentMap.vue_calculate_moment(self, *args)
    227 def vue_calculate_moment(self, *args):
--> 228     self.calculate_moment(add_data=True)

File ~\Documents\STScI\jdaviz\jdaviz\core\template_mixin.py:294, in with_spinner.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
    292 setattr(self, spinner_traitlet, True)
    293 try:
--> 294     ret_ = meth(self, *args, **kwargs)
    295 except Exception:
    296     setattr(self, spinner_traitlet, False)

File ~\Documents\STScI\jdaviz\jdaviz\configs\cubeviz\plugins\moment_maps\moment_maps.py:177, in MomentMap.calculate_moment(self, add_data)
    175         cube = self.dataset.selected_obj
    176 else:
--> 177     _, _, cube = self._get_continuum(self.dataset,
    178                                      'per-pixel',
    179                                      self.spectral_subset,
    180                                      update_marks=False)
    182 # slice out desired region
    183 # TODO: should we add a warning for a composite spectral subset?
    184 spec_min, spec_max = self.spectral_subset.selected_min_max(cube)

File ~\Documents\STScI\jdaviz\jdaviz\core\template_mixin.py:2087, in SpectralContinuumMixin._get_continuum(self, dataset, spatial_subset, spectral_subset, update_marks)
   2083     sr = self.app.get_subsets(spectral_subset.selected,
   2084                               simplify_spectral=True,
   2085                               use_display_units=True)
   2086     spectrum = extract_region(full_spectrum, sr, return_single_spectrum=True)
-> 2087     sr_lower = np.nanmin(spectrum.spectral_axis[spectrum.spectral_axis.value >= sr.lower.value])  # noqa
   2088     sr_upper = np.nanmax(spectrum.spectral_axis[spectrum.spectral_axis.value <= sr.upper.value])  # noqa
   2090 if self.continuum_subset_selected == 'None':

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\astropy\units\quantity.py:1838, in Quantity.__array_function__(self, function, types, args, kwargs)
   1825 # A function should be in one of the following sets or dicts:
   1826 # 1. SUBCLASS_SAFE_FUNCTIONS (set), if the numpy implementation
   1827 #    supports Quantity; we pass on to ndarray.__array_function__.
   (...)
   1835 # function is in none of the above, we simply call the numpy
   1836 # implementation.
   1837 if function in SUBCLASS_SAFE_FUNCTIONS:
-> 1838     return super().__array_function__(function, types, args, kwargs)
   1840 elif function in FUNCTION_HELPERS:
   1841     function_helper = FUNCTION_HELPERS[function]

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\numpy\lib\nanfunctions.py:350, in nanmin(a, axis, out, keepdims, initial, where)
    347 else:
    348     # Slow, but safe for subclasses of ndarray
    349     a, mask = _replace_nan(a, +np.inf)
--> 350     res = np.amin(a, axis=axis, out=out, **kwargs)
    351     if mask is None:
    352         return res

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\astropy\units\quantity.py:1838, in Quantity.__array_function__(self, function, types, args, kwargs)
   1825 # A function should be in one of the following sets or dicts:
   1826 # 1. SUBCLASS_SAFE_FUNCTIONS (set), if the numpy implementation
   1827 #    supports Quantity; we pass on to ndarray.__array_function__.
   (...)
   1835 # function is in none of the above, we simply call the numpy
   1836 # implementation.
   1837 if function in SUBCLASS_SAFE_FUNCTIONS:
-> 1838     return super().__array_function__(function, types, args, kwargs)
   1840 elif function in FUNCTION_HELPERS:
   1841     function_helper = FUNCTION_HELPERS[function]

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\numpy\core\fromnumeric.py:2970, in amin(a, axis, out, keepdims, initial, where)
   2957 @array_function_dispatch(_min_dispatcher)
   2958 def amin(a, axis=None, out=None, keepdims=np._NoValue, initial=np._NoValue,
   2959          where=np._NoValue):
   2960     """
   2961     Return the minimum of an array or minimum along an axis.
   2962 
   (...)
   2968     ndarray.min : equivalent method
   2969     """
-> 2970     return _wrapreduction(a, np.minimum, 'min', axis, None, out,
   2971                           keepdims=keepdims, initial=initial, where=where)

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\numpy\core\fromnumeric.py:86, in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     84             return reduction(axis=axis, dtype=dtype, out=out, **passkwargs)
     85         else:
---> 86             return reduction(axis=axis, out=out, **passkwargs)
     88 return ufunc.reduce(obj, axis, dtype, out, **passkwargs)

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\numpy\core\_methods.py:45, in _amin(a, axis, out, keepdims, initial, where)
     43 def _amin(a, axis=None, out=None, keepdims=False,
     44           initial=_NoValue, where=True):
---> 45     return umr_minimum(a, axis, None, out, keepdims, initial, where)

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\astropy\coordinates\spectral_quantity.py:87, in SpectralQuantity.__array_ufunc__(self, function, method, *inputs, **kwargs)
     85 def __array_ufunc__(self, function, method, *inputs, **kwargs):
     86     # We always return Quantity except in a few specific cases
---> 87     result = super().__array_ufunc__(function, method, *inputs, **kwargs)
     88     if (
     89         (
     90             function is np.multiply
   (...)
     98         )
     99     ):
    100         result = result.view(self.__class__)

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\astropy\units\quantity.py:691, in Quantity.__array_ufunc__(self, function, method, *inputs, **kwargs)
    689     return NotImplemented
    690 else:
--> 691     raise e

File ~\anaconda3\envs\test-mm-sub\lib\site-packages\astropy\units\quantity.py:666, in Quantity.__array_ufunc__(self, function, method, *inputs, **kwargs)
    663     arrays.append(converter(input_) if converter else input_)
    665 # Call our superclass's __array_ufunc__
--> 666 result = super().__array_ufunc__(function, method, *arrays, **kwargs)
    667 # If unit is None, a plain array is expected (e.g., comparisons), which
    668 # means we're done.
    669 # We're also done if the result was None (for method 'at') or
    670 # NotImplemented, which can happen if other inputs/outputs override
    671 # __array_ufunc__; hopefully, they can then deal with us.
    672 if unit is None or result is None or result is NotImplemented:

ValueError: zero-size array to reduction operation minimum which has no identity

```

</details>

**Comment by kecnry** [comment]
What is your spectral subset?  Is it possible it doesn't encompass any data (in which case we should check for that - but probably somewhere more general).

**Comment by javerbukh** [comment]
Still seeing this in a new environment while changing where I make the spectral subset selection. Here is my environment:
```
Windows-10-10.0.19045-SP0
Python 3.10.13 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:24:38) [MSC v.1916 64 bit (AMD64)]
Numpy 1.26.2
astropy 6.0.0
matplotlib 3.8.2
scipy 1.11.4
scikit-image 0.22.0
asdf 3.0.1
stdatamodels 1.8.4
gwcs 0.20.0
regions 0.8
specutils 1.12.0
specreduce 1.3.0
photutils 1.10.0
astroquery 0.4.6
pyyaml 6.0.1
asteval 0.9.31
idna 3.6
traitlets 5.14.0
bqplot 0.12.42
bqplot-image-gl 1.4.11
glue-core 1.17.1
glue-jupyter 0.20.0
glue-astronomy 0.10.0
echo 0.8.0
ipyvue 1.10.1
ipyvuetify 1.8.10
ipysplitpanes 0.2.0
ipygoldenlayout 0.4.0
ipypopout 1.2.0
Jinja2 3.1.2
voila 0.4.3
vispy 0.14.1
sidecar 0.7.0
Jdaviz 3.8.1.dev19+g9b6350ff.d20231211
```

File: jdaviz/core/template_mixin.py [hunk 1]
```diff
@@ -1979,6 +1987,217 @@ def _check_non_finite_uncertainty_mismatch(self, event={}):
...
+    def _get_continuum(self, dataset, spatial_subset, spectral_subset, update_marks=False):
+        if dataset.selected == '':
+            self._update_continuum_marks()
+            return None, None, None
+
+        if spatial_subset == 'per-pixel':
+            if self.app.config != 'cubeviz':
+                raise ValueError("per-pixel only supported for cubeviz")
+            if "_orig_spec" in self.dataset.selected_obj.meta:
+                full_spectrum = self.dataset.selected_obj.meta["_orig_spec"]
```
Comment (javerbukh) [inline jdaviz/core/template_mixin.py:2057]: @kecnry The bug I'm seeing is triggered by this line. If I use the `full_spectrum` from line 2059 in a notebook, I see a result. If I use this line, I see the traceback I mentioned here https://github.com/spacetelescope/jdaviz/pull/2587#issuecomment-1846099676 .

Comment (kecnry) [reply jdaviz/core/template_mixin.py:2057]: curious, ok, I'll revert that logic and we can re-introduce it if we run into a case where its necessary.

**Review by javerbukh — APPROVED** [review]
The plugin works well and everything behaves as intended, although I'll let someone with more knowledge of continuum subtraction check the results of the tests. Nice work!

**Comment by kecnry** [comment]
> the green light

ha! 🤣 🟢

**Pull request merged**